src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -1221,3 +1288,31 @@ int sd_dhcp_server_set_callback(sd_dhcp_server *server, sd_dhcp_server_callback_
 
         return 0;
 }
+
+int sd_dhcp_server_add_static_lease(sd_dhcp_server *server, sd_dhcp_static_lease *v) {
+        _cleanup_(sd_dhcp_static_lease_unrefp) sd_dhcp_static_lease *old = NULL;
+        int r;
+
+        assert_return(server, -EINVAL);
+        assert_return(v, -EINVAL);
+
+        r = hashmap_ensure_allocated(&server->static_leases_by_client_id, &dhcp_static_lease_hash_ops);
+        if (r < 0)
+                return r;
+
+        /* Static lease with an empty or omitted address is a valid entry,
+         * it removes any static lease with the specified mac address. */
+        if (v->address == 0) {
+                old = hashmap_remove(server->static_leases_by_client_id, &v->client_id);
+                return 0;
+        }
+
+        old = hashmap_get(server->static_leases_by_client_id, &v->client_id);
+        r = hashmap_replace(server->static_leases_by_client_id, &v->client_id, v);
```

Comment:
if this fails, then the hashmap will point to an invalidated entry, since you assigned `old` already, which has a cleanup function attached, which will run in the error path freeing an object that then still is part of the hashmap